src/extension/ip_extraction.c

```diff
@@ -0,0 +1,596 @@
+    _true_client_ip_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_TRUE_CLIENT_IP"), 1);
+    _remote_addr_key = zend_string_init_interned(ZEND_STRL("REMOTE_ADDR"), 1);
+
+    _register_testing_objects();
+}
+
+static zend_string *nullable _fetch_arr_str(
+    const zval *nonnull server, zend_string *nonnull key);
+static bool _is_private(const ipaddr *nonnull addr);
+static zend_string *nullable _ipaddr_to_zstr(const ipaddr *ipaddr);
+static zend_string *_try_extract(
+    const zval *nonnull server, zend_string *key, extract_func_t extract_func);
+static bool _parse_x_forwarded_for(
+    zend_string *nonnull value, ipaddr *nonnull out);
+static bool _parse_plain(zend_string *nonnull zvalue, ipaddr *nonnull out);
+static bool _parse_forwarded(zend_string *nonnull zvalue, ipaddr *nonnull out);
+static bool _parse_via(zend_string *nonnull zvalue, ipaddr *nonnull out);
+
+zend_string *nullable dd_ip_extraction_find(zval *nonnull server)
+{
+    zend_string *res;
+
+    if (_ipheader) {
+        zend_string *value = _fetch_arr_str(server, _ipheader);
+        if (!value) {
+            return NULL;
+        }
+
+        ipaddr out;
+        bool succ;
+        succ = _parse_forwarded(value, &out);
+        if (succ) {
+            return _ipaddr_to_zstr(&out);
+        }
+
+        succ = _parse_x_forwarded_for(value, &out);
+        if (succ) {
+            return _ipaddr_to_zstr(&out);
+        }
+
+        return NULL;
+    }
+
+    res = _try_extract(server, _x_forwarded_for_key, &_parse_x_forwarded_for);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _x_real_ip_key, &_parse_plain);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _client_ip_key, &_parse_plain);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _x_forwarded_key, &_parse_forwarded);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _x_cluster_client_ip_key, &_parse_plain);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _forwarded_for_key, &_parse_x_forwarded_for);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _forwarded_key, &_parse_forwarded);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _via_key, &_parse_via);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _true_client_ip_key, &_parse_plain);
+    if (res) {
+        return res;
+    }
+
+    res = _try_extract(server, _remote_addr_key, &_parse_plain);
+    if (res) {
+        return res;
+    }
+
+    return NULL;
+}
+
+static zend_string *_try_extract(
+    const zval *nonnull server, zend_string *key, extract_func_t extract_func)
+{
+    zend_string *value = _fetch_arr_str(server, key);
+    if (!value) {
+        return NULL;
+    }
+    ipaddr out;
+    if ((*extract_func)(value, &out)) {
+        if (_is_private(&out)) {
```

Comment:
This seems to be duplicate work for all extraction functions except the plain one, perhaps you should check if the IP is private in that function as well and remove this from here.